cmd/neofs-cli/modules/root.go

```diff
@@ -183,29 +173,40 @@ func getKey() (*ecdsa.PrivateKey, error) {
 		return &priv.PrivateKey, nil
 	}
 
-	if keyPath := viper.GetString(binaryKey); keyPath != "" {
-		return getKeyFromFile(keyPath)
+	// Ideally we want to touch file-system on the last step.
+	// However, asking for NEP-2 password seems to be confusing if we provide a wallet.
+	// Thus we try keys in the following order:
+	// 1. WIF
+	// 2. Raw binary key
+	// 3. Wallet file
+	// 4. NEP-2 encrypted WIF.
+	keyDesc := viper.GetString(walletPath)
+	priv, err := keys.NewPrivateKeyFromWIF(keyDesc)
+	if err == nil {
+		return &priv.PrivateKey, nil
 	}
 
-	if walletPath := viper.GetString(walletPath); walletPath != "" {
-		w, err := wallet.NewWalletFromFile(walletPath)
-		if err != nil {
-			return nil, fmt.Errorf("%w: %v", errInvalidKey, err)
-		}
-		return getKeyFromWallet(w, viper.GetString(address))
+	p, err := getKeyFromFile(keyDesc)
+	if err == nil {
+		return p, nil
 	}
 
-	wif := viper.GetString(wif)
-	if len(wif) == nep2Base58Length {
-		return getKeyFromNEP2(wif)
+	w, err := wallet.NewWalletFromFile(keyDesc)
+	if err == nil {
+		return getKeyFromWallet(w, viper.GetString(address))
 	}
 
-	priv, err := keys.NewPrivateKeyFromWIF(wif)
-	if err != nil {
-		return nil, fmt.Errorf("%w: %v", errInvalidKey, err)
+	if len(keyDesc) == nep2Base58Length {
+		// Check base58-encoding to avoid returning invalid password error when key is in fact incorrect.
+		for _, c := range keyDesc {
+			if !strings.ContainsRune("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz", c) {
+				return nil, errInvalidKey
```

Comment:
wouldn't it be better to have a special error for that case? "provided key is incorrect" may confuse I think especially if I use WIF with `wallet` flag

Reply:
I have removed this part. Checking full NEP-2 format seems an overkill, and checking length should be enough for most cases.